impl/src/main/kotlin/nl/nuts/consent/bridge/corda/CordaManagedConnection.kt

```diff
@@ -40,60 +41,113 @@ import java.net.Socket
+                connection?.forceClose()
+                connection = null
+            }
+        }
+
+        override fun run() {
+            try {
+                while (!terminate) {
+                    synchronized(this) {
+                        if (connection != null) {
+                            try {
+                                val nodeInfo = connection?.proxy?.nodeInfo()
+                                require(nodeInfo?.legalIdentitiesAndCerts?.isNotEmpty() ?: false)
+                            } catch (e: Exception) {
+                                when (e) {
+                                    is IllegalArgumentException, is ConnectionFailureException, is RPCException -> {
+                                        noConnectionReason = e.message
+                                        logger.error("Corda RPC connection lost for $managedConnection.name: ${e.message}")
+                                        connection = null
+                                        onDisconnected()
+                                    }
+                                    else -> throw e
+                                }
+                            }
+                        } else if (tryToConnect) {
+                            try {
+                                noConnectionReason = null
+                                connection = managedConnection.connectToCorda()
+                                logger.info("Corda RPC connection established for ${managedConnection.name}")
+                                onConnected()
+                            } catch (e: Exception) {
+                                when (e) {
+                                    is IOException, is ActiveMQSecurityException, is RPCException -> {
+                                        onError()
+                                        logger.debug("${e::class.simpleName}: ${e.message} while trying to connect to Corda for ${managedConnection.name}")
+                                        noConnectionReason = e.message
+                                    }
+                                    else -> throw e
+                                }
+                            }
                         }
                     }
+                    Thread.sleep(delay)
                 }
-                Thread.sleep(consentBridgeRPCProperties.retryIntervalSeconds.seconds.toMillis())
+            } catch (e: Exception) {
+                logger.error("unexpected exception in watchdog for ${managedConnection.name}: ${e.message}", e)
             }
         }
-        watchdog?.start()
+    }
+
+    fun noConnectionReason(): String? {
+        return cordaConnectionWatchdog?.noConnectionReason
+    }
+
+    private fun startThread() {
+        cordaConnectionWatchdog = CordaConnectionWatchdog(this, consentBridgeRPCProperties.retryIntervalSeconds.seconds.toMillis(), this.onConnected, this.onDisconnected)
+        Thread(cordaConnectionWatchdog).start()
```

Comment:
maybe mark this as daemon thread to avoid it blocking JVM shutdown in abnormal situations?

Reply:
done